### PR TITLE
Bind the edit program API with the view

### LINKED ISSUE #26: Bind the edit program view with the API
**Is your feature request related to a problem? Please describe.**
- We need to bind the edit program view with the API so that we can edit a program

**Describe the solution you'd like**
- Bind the edit program view with the relevant API

### LINKED ISSUE #27: Remove duplicated program interfaces
**Is your feature request related to a problem? Please describe.**
The `Program` interface has been duplicated in two places. `/src/interfaces.ts` and `src/scenes/Home/components/AddProgram/interfaces.ts` 
The reason for having two interfaces is, the `Program` in Add Program view doesn't have the `id` property while it is a required property for the saved programs. 

**Describe the solution you'd like**
Implement the interfaces using inheritance.

### PR DESCRIPTION
**Pull request by anjula-sack**
## Purpose

- The purpose of this PR is to fix #26 and to fix #27 

## Goals

- Edit program details using the dashboard

## Approach

- Bound the API with the edit program view

## Screenshots

Kapture 2020-10-24 at 10 18 24

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx [hunk 1]
```diff
@@ -1,53 +1,122 @@
...
 const { Title } = Typography;
 
-function EditProgram() {
+function EditProgram(props: RouteComponentProps<ProgramUrlParams>) {
   const [form] = Form.useForm();
+  const [programTitle, setProgramTitle] = useState<string>('');
+  const [programHeading, setProgramHeading] = useState<string>('');
+  const [programImageURL, setProgramImageURL] = useState<string>('');
+  const [programLandingURL, setProgramLandingURL] = useState<string>('');
+  const [isLoading, setIsLoading] = useState<boolean>(false);
```
Comment (jayasanka-sack) [inline src/scenes/Dashboard/scenes/EditProgram/index.tsx]: Is there a requirement that the component should get rerendered every time we change some detail of the program? If no, definitely it should not be in the state

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx [hunk 2]
```diff
@@ -1,53 +1,122 @@
...
+
+  const programId: string = props.match.params.programId;
```
Comment (jayasanka-sack) [inline src/scenes/Dashboard/scenes/EditProgram/index.tsx]: Use the router hook

Comment (anjula-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: the what?

Comment (jayasanka-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: To get the path param.

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx [hunk 3]
```diff
@@ -1,53 +1,122 @@
...
 
+  form.setFieldsValue({
+    title: programTitle,
+    imgURL: programImageURL,
+    landingURL: programLandingURL,
+    heading: programHeading,
+  });
+
+  const addProgram = (values: any) => {
+    const program = {
```
Comment (jayasanka-sack) [inline src/scenes/Dashboard/scenes/EditProgram/index.tsx]: data type

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx [hunk 4]
```diff
@@ -1,53 +1,122 @@
...
+            Save
+          </Button>
+        </Form>
+      </Spin>
     </div>
   );
 }
 
-export default EditProgram;
+export default withRouter(EditProgram);
```
Comment (jayasanka-sack) [inline src/scenes/Dashboard/scenes/EditProgram/index.tsx]: HOCs are for the class components. Hooks are for the functional components. 

https://sophieau.com/article/hoc-renderprop-hook/

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx (hunk 2 above)
Comment (anjula-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: yeah got it

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx (hunk 1 above)
Comment (anjula-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: yeah

Comment (jayasanka-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: Not fixed this yet. You can do this without using state.

**Comment by jayasanka-sack** [comment]
Please fix the #27 with this pr too

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx [hunk 5]
```diff
@@ -1,51 +1,128 @@
...
+          message: 'Warning!',
+          description: 'Something went wrong when fetching the program',
+        });
+      });
   }, []);
 
+  form.setFieldsValue({
+    title: program.title,
+    imgURL: program.imageUrl,
+    landingURL: program.landingPageUrl,
```
Comment (jayasanka-sack) [inline src/scenes/Dashboard/scenes/EditProgram/index.tsx]: Move this to the success callback of the fetch program.
Then we can completely move the program out of the state

**Comment by jayasanka-sack** [comment]
> Please fix the #27 with this pr too

No pressure. Only if you could

File: src/scenes/Dashboard/scenes/EditProgram/index.tsx (hunk 5 above)
Comment (anjula-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: do you mean not to use the state?

Comment (anjula-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: how are we going to re-render the form with changed values if we didn’t  use the state

Comment (jayasanka-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: I think we can use the form.setFieldValues hook

Comment (jayasanka-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: from the Antd’s form API page,

After wrapped by Form.Item with name property, value(or other property defined by valuePropName) onChange(or other property defined by trigger) props will be added to form controls, the flow of form data will be handled by Form which will cause:

> 

You shouldn't use onChange on each form control to collect data(use onValuesChange of Form), but you can still listen to onChange.
You cannot set value for each form control via value or defaultValue prop, you should set default value with initialValues of Form. Note that initialValues cannot be updated by setState dynamiclly, you should use setFieldsValue in that situation.
You shouldn't call setState manually, please use form.setFieldsValue to change value programmatically.

>

Comment (anjula-sack) [reply src/scenes/Dashboard/scenes/EditProgram/index.tsx]: yes

**Review by jayasanka-sack — APPROVED** [review]
thanks @anjula-sack !

**Pull request merged**